[WIP] Fix for timeout in graph_model

rpgoldman opened the pull request:
See #3458

rpgoldman committed:
Revise _get_ancestors to use BFS.

Previously, `get_ancestors()` used a powerset computation, which would fail on large models. Replaced with breadth-first search.

rpgoldman commented:
This PR is *almost* right, I think, but somehow by changing the ancestor computation, I have changed the computation of the dimensions in graph rendering.

lucianopaz commented on pymc3/model_graph.py:
```diff
@@ -40,24 +47,38 @@ def _ancestors(self, var, func, blockers=None):
+        vars: List[var] = set(self.var_list)
+        vars.remove(var)
+        
         upstream = self._ancestors(var, func)
```
I would completely remove `self._ancestors` and replace it with the `stack_search` that you did

lucianopaz reviewed (changes requested):
Given that you observed a change in the computed dimensions, could you add a new test to the test suite that checks if the computed dimensions match with what should be returned?

rpgoldman commented:
@lucianopaz actually the dimension change was not an issue in the breadth first search -- it revealed an error in the code I used to debug the search! That's a relief. I will see if I can propose some tests, but I don't understand the test suite  very well.
I will try to make the change you suggest today.

rpgoldman committed:
Simplify _get_ancestors().

Per @lucianopaz. remove the initial special case/

rpgoldman commented:
@lucianopaz I have just looked over the existing tests for the model grapher, and they check the structure, including the identification of plates.  So I think they are sufficient in their present form.  There is no obvious test to add based on the fix, because the old version didn't compute the wrong thing; it simply didn't complete. But I am open to suggestions.

Note: if this PR looks good, I should squash it before it's merged.

lucianopaz commented on pymc3/model_graph.py:
```diff
@@ -40,24 +47,32 @@ def _ancestors(self, var, func, blockers=None):
         return set([j for j in ancestors([func], blockers=blockers) if j in self.var_list and j != var])
```
`_ancestors` now seems to be unused, right? If so, it should be removed.

lucianopaz reviewed (changes requested):
Everything looks fine to me. No other tests should be needed. Once `_ancestors` is removed, just add a line to the release notes under maintenance and the PR should be ready to merge

rpgoldman committed:
Remove unused _ancestors method

New Breadth-first search version of _get_ancestors no longer uses it.
Also added some type annotations to clarify.

rpgoldman commented:
@lucianopaz OK, I think this is done now.

lucianopaz commented:
Great! Thanks a lot @rpgoldman! Once the tests pass, I'll merge

lucianopaz reviewed (changes requested):
Oh, I had forgotten that you still need to add a line to RELEASE-NOTES.md under the maintenance section referencing the fixed issue

rpgoldman committed:
Release note for PR #3460

rpgoldman commented:
@lucianopaz Pushed the documentation and removed an unused import causing the check failure.

lucianopaz commented:
Thanks again @rpgoldman!

rpgoldman commented:
@lucianopaz No problem! I'm glad to be able to give even a little something back in exchange for all I've gotten from the community!

twiecki commented:
Thanks @rpgoldman!
